include/utils/types/geometry.h

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2023 UltiMaker
+// CuraEngine is released under the terms of the AGPLv3 or higher
+
+/*! @file geometry.h
+ *  @brief Defines geometric concepts like Point2D, Point3D, Segment etc.
+ */
+
+#ifndef UTILS_TYPES_GEOMETRY_H
+#define UTILS_TYPES_GEOMETRY_H
+
+#include <concepts>
+#include <string>
+#include <tuple>
+#include <type_traits>
+
+#include <range/v3/range/concepts.hpp>
+#include <range/v3/range/operations.hpp>
+
+#include "utils/types/generic.h"
+
+namespace cura::utils
+{
+
+/*!
+ * @concept point2d_tuple
+ * @brief Checks whether T is a 2D point represented as a tuple with two integral elements
+ * @tparam T The type to check
+ */
+template<typename T>
+concept point2d_tuple = requires(T t)
+{
+    requires std::is_same_v<T, std::tuple<typename std::tuple_element<0, T>::type, typename std::tuple_element<0, T>::type>>;
```

Comment:
```suggestion
    requires std::is_same_v<T, std::tuple<typename std::tuple_element<0, T>::type, typename std::tuple_element<1, T>::type>>;
```

Reply:
This is intentional, this forces to have both elements the same type